File: app/modules/unionBerlinTickets/unionBerlinTickets.go
package union

import (
	"log"
	"main/utils"
	"strings"

	"github.com/PuerkitoBio/goquery"
)

const unionBerlinUrl = "https://tickets.union-zeughaus.de/unveu/heimspiele_2.htm"

func UnionBerlinTickets(dynamodb *utils.DynamoDbService) *UnionBerlinTicketsData {
	res, err := unionBerlinTickets()
	if err != nil {
		log.Printf("Error in Union Berlin Tickets module: %s", err)
	}
	res.data = dynamodb.GetValueIfChanged("union_berlin_tickets", res.data)
	return res
}

func unionBerlinTickets() (*UnionBerlinTicketsData, error) {
	doc, err := utils.GetDoc(unionBerlinUrl)
	if err != nil {
		return nil, err
	}

	tickets := ""
	parentDiv := doc.Find(".ticket.listitem.gamehome")
	parentDiv.Each(func(i int, s *goquery.Selection) {
		elementsStrings := []string{}
		s.Find("h2").Each(func(i int, s *goquery.Selection) {
			elementsStrings = append(elementsStrings, s.Text())
		})
		tickets += strings.Join(elementsStrings, " - ") + "\n"
	})

	return &UnionBerlinTicketsData{tickets}, nil
}

File: modules/unionBerlinTickets.py
import requests
from bs4 import BeautifulSoup
import logging

from utils.dynamodb import ItemStoreService


def __union_berlin_tickets():
    union_berlin_tickets_url = "https://tickets.union-zeughaus.de/unveu/heimspiele_2.htm"
    res = f"[Union Berlin tickets]({union_berlin_tickets_url}):\n"
    headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:82.0) Gecko/20100101 Firefox/82.0'}
    html_text = requests.get(union_berlin_tickets_url, timeout=5, headers=headers).text
    soup = BeautifulSoup(html_text, 'html.parser')

    for event in soup.findAll(True, {"class": ['ticket listitem gamehome']}):
        res += " - ".join([team.text for team in event.find_all_next("h2")])
        res += "\n"

    return res


def union_berlin_tickets(item_store_service: ItemStoreService) -> str:
    try:
        actual_tickets = __union_berlin_tickets()
        if item_store_service is None:
            return actual_tickets
        stored_tickets = item_store_service.get_item("union_berlin_tickets")
        if actual_tickets != stored_tickets:
            item_store_service.save_item("union_berlin_tickets", actual_tickets)
            return actual_tickets
        else:
            return ""
    except:
        logging.exception("Couldn't get Union Berlin tickets")
        return "Couldn't get Union Berlin tickets"
